### DISCUSSION TITLE
About backlinks

### DISCUSSION THREAD

**Discussion by bepolymathe** [freeplane/freeplane#485] [General]
Hi all !

I find that one of the interesting recent developments in note-taking software in recent years has been the integration of backlinks. I'm not going to leave Freeplane just for this feature but I would like to have it in my FP. Have any of you developed uses or thoughts around this?
Nice day to you.

**Comment by macmarrum** [comment]
I am not sure what you mean by "the integration of backlinks". Could you please give an example?

**Comment by bepolymathe** [comment]
Yes of course. You could say that "backlinks" are a way to quickly see which nodes are connected to the node being edited. 

Here an example for Obsidian. 
https://jackiegeek.gitee.io/obsidian-docs/fr/How%20to/Working%20with%20Backlinks/

The advantage is to be able to connect ideas (nodes?) in a non-hierarchical way.

**Reply by macmarrum** [reply]
I see.

There are several ways of linking a node to other nodes, in addition to the parent-child relationship.

1. A node can have a link to another: `Insert->Link->Add local hyperlink`
2. A node can have several links to other nodes, each as a separate attribute: `Edit->Node properties->Edit attribute in-line`, add an attribute (right-click if needed) then right-click the value and `Set Node Link...`
3. A node can have one or more Connectors to other nodes: https://docs.freeplane.org/attic/#/wiki/understanding-nodes?id=connectors – this is probably the functionality which most resembles your description of "backlinks"
BTW: Connector lines can be hidden: `Preferences…->Defaults->Connectors->Paint connectors behind nodes`
Like for clones, the right-click node menu item `Goto` list the other ends of Connectors.

**Reply by macmarrum** [reply]
I just realize, there's one more:

4. Core text, Note or Details can contain (HTML) hyperlinks to other nodes

**Reply by bepolymathe** [reply]
> 3\. A node can have one or more Connectors to other nodes: https://docs.freeplane.org/attic/#/wiki/understanding-nodes?id=connectors – this is probably the functionality which most resembles your description of "backlinks"

Thanks for reply. Indeed it opens possibilities but I don't know if it's so practical. I'll explain you a use case. I would like to create a node for all the people with whom I have regular exchanges and link all the notes that concern them. When I search for the nodes that concern this person I only have to go to the person node to have them listed. I don't know if I can do this quickly... 

In this case you might as well make a classic attribute with "person" and "First name Last name" and search through the filters.

**Reply by macmarrum** [reply]
> When I search for the nodes that concern this person I only have to go to the person node to have them listed. I don't know if I can do this quickly...

You could use clones of the nodes that concern a person, located in a "Backlinks" node for that person, e.g.

```
Person A
  Backlinks
    Node A (a clone)
    Node B (a clone)
```

or, even better, use link-nodes (aka symlinks) in place of clones. This script might be helpful: https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/pasteAsSymlink.groovy

**Reply by bepolymathe** [reply]
> You could use clones of the nodes that concern a person, located in a "Backlinks" node for that person, e.g.

Yes, but this requires a manual operation for each link creation. So I think it's better with attributes. 

> or, even better, use link-nodes (aka symlinks) in place of clones

It could be interesting! Thanks for the script, I'll think about it...

**Reply by ZbyszekS** [reply]
I copied script https://github.com/macmarrum/freeplane-scripting/blob/main/src/scripts/pasteAsSymlink.groovy to script/init folder and restart freeplane. Got in menu Mac1>Paste>Past As Symlink but clicking it gives no effect on the node

The ideal solution would be:
1. You choose, from menu or through shortcut to create "Backlinks" for current Node
2. You got smth like: 
current_Node
  Backlinks
    Node A (a clone)
    Node B (a clone)

**Comment by macmarrum** [comment]
@ZbyszekS 
> Got in menu `Mac1>Paste>Past As Symlink` but clicking it gives no effect on the node

Is this what you mean? You copy a node (or several nodes) with `Ctrl+C`, then run `Mac1->Paste->Past As Symlink` but nothing happens?

**Reply by ZbyszekS** [reply]
Now I get it. I was so focused on searching what functionality I want, that I just overlooked how it works. And yes, it works and places links to designed node, as it should be.

**Comment by macmarrum** [comment]
> The ideal solution would be:
>
> You choose, from menu or through shortcut to create "Backlinks" for current Node
You got smth like:
current_Node
Backlinks
Node A (a clone)
Node B (a clone)

I see.
Could you please talk some more about how you link your nodes?
I counted [4 ways](https://github.com/freeplane/freeplane/discussions/485#discussioncomment-3077427) that can be used in Freeplane. Which ones do you use?

**Reply by ZbyszekS** [reply]
90% of the time I link nodes locally through / as "Insert>Link>Add local hyperlink"  (alt-shift-l). Sometimes I link nodes through anchor mechanism. This by ideal solution I meant the script/subprogram that could "read" all nodes "pointing" to current/selected node, and based on that knowledge, would create subnode, to current node, called for example "Backlinks", and there place sym(links) to all these nodes pointing towards current node.

**Reply by euu2021** [reply]
> 90% of the time I link nodes locally through / as "Insert>Link>Add local hyperlink" (alt-shift-l). Sometimes I link nodes through anchor mechanism. This by ideal solution I meant the script/subprogram that could "read" all nodes "pointing" to current/selected node, and based on that knowledge, would create subnode, to current node, called for example "Backlinks", and there place sym(links) to all these nodes pointing towards current node.

This would be an interesting usecase for an [Utility Panel](https://github.com/euu2021/Freeplane_UtilityPanels). So, when the user selects a node, the panel shows a list of all nodes that point to that node. If the user selects multiple nodes, the list would be a sum of all the backlinks.

I will include this in the TODO list.

**Reply by euu2021** [reply]
Also, it would be nice to have a script that automatically creates a tag in the node with the number of backlinks that node has. So, when it's created a link to that node, the number is updated.
